### PR TITLE
CAN Transceiver Flow

### LINKED ISSUE #66: CAN Transceiver Full Sequence Skeleton
### Purpose

Lay the foundation for the CAN Transceiver data flows seen here: https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1768914974/CAN+Transceiver+Design+Specification#Control-Sequences.

### Description

Adding skeleton data flows provide useful context when implementing the actual logic and processing.

### Resources

- For callback functions, you can use either [std::function](https://en.cppreference.com/w/cpp/utility/functional/function) or [function pointers](https://www.cprogramming.com/tutorial/function-pointers.html)
  - If using function pointers, alias the pointers with `using`: https://stackoverflow.com/a/16498976 (ignore the bit about making it "less ugly")

### PR DESCRIPTION
**Pull request by hhenry01**
### Description

- Starter code for #87 
- Includes CAN Parsing interface and an implementation for the Battery
- Closes #66 

### Verification

- Unit tested.
- ROS launch works as expected with subscribed and published topics.

### TODO
Once merged, create an issue to route AIS data through the CAN transceiver. Currently, the CAN Transceiver ROS Intf just passes the mock AIS directly from the MOCK_AIS_TOPIC to the AIS_TOPIC. This is because ELEC has not defined AIS CAN frame formats.

### Resources

-

### CODE REVIEW

File: projects/can_transceiver/src/can_transceiver_ros_intf.cpp [hunk 1]
```diff
@@ -6,236 +6,190 @@
...
 
 #include "can_frame_parser.h"
 #include "can_transceiver.h"
 #include "cmn_hdrs/ros_info.h"
+#include "cmn_hdrs/shared_constants.h"
 
-constexpr int QUEUE_SIZE        = 10;
-constexpr int PLACEHOLDER_VALUE = 42;   // Placeholder value for debugging or testing
-constexpr int PLACEHOLDER_MS    = 500;  // Timer to be replaced with callback in future
+constexpr int  QUEUE_SIZE     = 10;
```
[the remark this answers is not in the document]
Comment (hhenry01) [reply projects/can_transceiver/src/can_transceiver_ros_intf.cpp:19]: Arbitrary. I'll add a comment.

File: projects/can_transceiver/src/can_transceiver_ros_intf.cpp [hunk 2]
```diff
@@ -6,236 +6,190 @@
...
     {
-        static constexpr int  ROS_Q_SIZE     = 5;
-        static constexpr auto TIMER_INTERVAL = std::chrono::milliseconds(500);
-        pub_   = this->create_publisher<std_msgs::msg::String>(PLACEHOLDER_TOPIC_0_TOPIC, ROS_Q_SIZE);
-        timer_ = this->create_wall_timer(TIMER_INTERVAL, std::bind(&CanTransceiverIntf::pub_cb, this));
-        sub_   = this->create_subscription<std_msgs::msg::String>(
-          PLACEHOLDER_TOPIC_1_TOPIC, ROS_Q_SIZE, std::bind(&CanTransceiverIntf::sub_cb, this, std::placeholders::_1));
+        this->declare_parameter("enabled", true);
+
+        if (!this->get_parameter("enabled").as_bool()) {
```
[the remark this answers is not in the document]
Comment (hhenry01) [reply projects/can_transceiver/src/can_transceiver_ros_intf.cpp:33]: That true is a default value.

[the remark this answers is not in the document]
Comment (hhenry01) [reply projects/can_transceiver/src/can_transceiver_ros_intf.cpp:33]: This is the resource I used to learn about ros parameters in code: https://roboticsbackend.com/rclcpp-params-tutorial-get-set-ros2-params-with-cpp/

File: projects/can_transceiver/src/can_transceiver.cpp [hunk 3]
```diff
@@ -15,102 +15,133 @@ using IFreq       = struct ifreq;
 using SockAddr    = struct sockaddr;
 using SockAddrCan = struct sockaddr_can;
 
-namespace
+using CAN_FP::CanFrame;
+using CAN_FP::CanId;
+
+void CanTransceiver::onNewCmd(const CanFrame & cmd_frame)
```
[the remark this answers is not in the document]
Comment (hhenry01) [reply projects/can_transceiver/src/can_transceiver.cpp:21]: * Now that you bring it up, it is a bit redundant to have both send() and onNewCmd() take the same inputs
* Taking a msg:: type is more hassle than it's worth. We would need to overload the function for *every* msg type just to call `.toLinuxCan()` and send() it

I will remove this function entirely.

**Review by samdai01 — APPROVED** [review]

**Pull request merged**